### PR TITLE
Moved static data endpoint to V3 and improved RequesterBase

### PR DESCRIPTION
**Pull request by JanOuborny**
Corresponding issue #352

### CODE REVIEW

**Comment by BenFradet** [comment]
Will review tomorrow.

File: RiotSharp/StaticRiotApi.cs [hunk 1]
```diff
@@ -24,46 +24,57 @@
...
-        private const string RuneCacheKey = "rune";
+        private const string RuneByIdCacheKey = "rune";
 
-        private const string SummonerSpellRootUrl = "/api/lol/static-data/{0}/v1.2/summoner-spell";
-        private const string SummonerSpellsCacheKey = "spells";
-        private const string SummonerSpellCacheKey = "spell";
+        private const string SummonerSpellsUrl = "summoner-spells";
+        private const string SummonerSpellsCacheKey = "summoner-spells";
+        private const string SummonerSpellByIdUrl = "summoner-spells/{0}";
+        private const string SummonerSpellCacheKey = "summoner-spell";
```
Comment (BenFradet) [inline RiotSharp/StaticRiotApi.cs]: I'd suggest renaming to `SummonSpellByIdCacheKey` to be consistent

File: RiotSharp/StaticRiotApi.cs [hunk 2]
```diff
@@ -789,11 +813,11 @@ public RealmStatic GetRealm(Region region)
             return wrapper.SummonerSpellListStatic;
         }
     
-        public SummonerSpellStatic GetSummonerSpell(Region region, SummonerSpell summonerSpell,
+        public SummonerSpellStatic GetSummonerSpell(Region region, int summonerSpellId,
             SummonerSpellData summonerSpellData = SummonerSpellData.basic, Language language = Language.en_US)
         {
             var wrapper = cache.Get<string, SummonerSpellStaticWrapper>(
-                SummonerSpellCacheKey + summonerSpell.ToCustomString());
+                SummonerSpellCacheKey + summonerSpellId.ToString());
```
Comment (BenFradet) [inline RiotSharp/StaticRiotApi.cs]: if you only keep the spell id as int, you don't need `ToString` when doing concatenation

File: RiotSharp/StaticRiotApi.cs [hunk 3]
```diff
@@ -827,18 +849,18 @@ public RealmStatic GetRealm(Region region)
                             string.Format("spellData={0}", summonerSpellData.ToString())
                         });
                     var spell = JsonConvert.DeserializeObject<SummonerSpellStatic>(json);
-                    cache.Add(SummonerSpellCacheKey + summonerSpell.ToCustomString(),
+                    cache.Add(SummonerSpellCacheKey + summonerSpellId.ToString(),
```
Comment (BenFradet) [inline RiotSharp/StaticRiotApi.cs]: same

File: RiotSharp/StaticRiotApi.cs [hunk 4]
```diff
@@ -827,18 +849,18 @@ public RealmStatic GetRealm(Region region)
...
             }
         }
       
-        public async Task<SummonerSpellStatic> GetSummonerSpellAsync(Region region, SummonerSpell summonerSpell,
+        public async Task<SummonerSpellStatic> GetSummonerSpellAsync(Region region, int summonerSpellId,
             SummonerSpellData summonerSpellData = SummonerSpellData.basic, Language language = Language.en_US)
         {
             var wrapper = cache.Get<string, SummonerSpellStaticWrapper>(
-                SummonerSpellCacheKey + summonerSpell.ToCustomString());
+                SummonerSpellCacheKey + summonerSpellId.ToString());
```
Comment (BenFradet) [inline RiotSharp/StaticRiotApi.cs]: same

File: RiotSharp/StaticRiotApi.cs [hunk 5]
```diff
@@ -863,11 +883,13 @@ public RealmStatic GetRealm(Region region)
                 });
             var spell = await Task.Factory.StartNew(() =>
                 JsonConvert.DeserializeObject<SummonerSpellStatic>(json));
-            cache.Add(SummonerSpellCacheKey + summonerSpell.ToCustomString(),
+            cache.Add(SummonerSpellCacheKey + summonerSpellId.ToString(),
```
Comment (BenFradet) [inline RiotSharp/StaticRiotApi.cs]: same

Comment (BenFradet) [inline RiotSharp/StaticRiotApi.cs]: also this should be renamed to `SummonerSpellByIdCacheKey`

File: RiotSharpTest/StaticRiotApiTest.cs [hunk 6]
```diff
@@ -191,6 +201,28 @@ public void GetMasteryAsync_Test()
             Assert.AreEqual(mastery.Result.Name, StaticRiotApiTestBase.staticMasteryName);
         }
 
+        #region Profile Icons Tests
+        [TestMethod]
+        [TestCategory("StaticRiotApi")]
+        public void GetProfileIcons_Test()
+        {
+            var profileIcons = api.GetChampions(StaticRiotApiTestBase.region, ChampionData.all);
```
Comment (BenFradet) [inline RiotSharpTest/StaticRiotApiTest.cs]: wrong api call

File: RiotSharpTest/StaticRiotApiTest.cs [hunk 7]
```diff
@@ -191,6 +201,28 @@ public void GetMasteryAsync_Test()
...
+
+            Assert.IsNotNull(profileIcons.Champions);
+            Assert.IsTrue(profileIcons.Champions.Count > 0);
+        }
+
+        [TestMethod]
+        [TestCategory("StaticRiotApi"), TestCategory("Async")]
+        public void GetProfileIconsAsync_Test()
+        {
+            var profileIcons = api.GetChampionsAsync(StaticRiotApiTestBase.region, ChampionData.all);
```
Comment (BenFradet) [inline RiotSharpTest/StaticRiotApiTest.cs]: same

File: RiotSharpTest/StaticRiotApiTest.cs [hunk 8]
```diff
@@ -254,7 +286,7 @@ public void GetSummonerSpellsAsync_Test()
         public void GetSummonerSpell_Test()
         {
             var spell = api.GetSummonerSpell(StaticRiotApiTestBase.region, 
-                StaticRiotApiTestBase.staticSummonerSpell, SummonerSpellData.all);
+                (int) StaticRiotApiTestBase.staticSummonerSpell, SummonerSpellData.all);
```
Comment (BenFradet) [inline RiotSharpTest/StaticRiotApiTest.cs]: are we sure this is a better API?

Comment (JanOuborny) [inline RiotSharpTest/StaticRiotApiTest.cs]: My reason behind this change is, that currently you get the summoner spell id as an int back (in the match endpoint). Therefore, it is more fluent to adjust the static api to also use an int.

File: RiotSharp/Http/Requester.cs [hunk 9]
```diff
@@ -17,8 +18,26 @@ public Requester(string apiKey) : base(apiKey)
...
+            return GetResponseContent(response);
+        }
+
+        public async Task<string> CreateGetRequestAsync(string relativeUrl, Region region,
+            List<string> addedArguments = null, bool useHttps = true)
+        {
+            rootDomain = GetPlatformDomain(region);
+            var request = PrepareRequest(relativeUrl, addedArguments, useHttps, HttpMethod.Get);
+            var response = GetAsync(request).Result;
+            return await GetResponseContentAsync(response);
```
Comment (JanOuborny) [inline RiotSharp/Http/Requester.cs]: I think we can just use an `await` here, but good catch. 👍

File: RiotSharpTest/StaticRiotApiTest.cs (hunk 8 above)
Comment (BenFradet) [inline RiotSharpTest/StaticRiotApiTest.cs]: Ok :+1: , then I think we can remove `ToCustomString`

File: RiotSharp/Interfaces/IStaticRiotApi.cs [hunk 10]
```diff
@@ -199,6 +200,22 @@ public interface IStaticRiotApi
             MasteryData masteryData = MasteryData.basic, Language language = Language.en_US);
 
         /// <summary>
+        /// Get a list of profile icons synchronously
+        /// </summary>
+        /// <param name="region"></param>
+        /// <param name="language"></param>
+        /// <returns></returns>
```
Comment (BenFradet) [inline RiotSharp/Interfaces/IStaticRiotApi.cs]: I don't think `param` and `returns` doc is needed if it's not filled

File: RiotSharp/StaticRiotApi.cs [hunk 11]
```diff
@@ -24,46 +24,57 @@
...
         private const string LanguagesCacheKey = "languages";
 
-        private const string MapRootUrl = "/api/lol/static-data/{0}/v1.2/map";
-        private const string MapCacheKey = "map";
+        private const string MapsUrl = "maps";
+        private const string MapsCacheKey = "maps";
 
-        private const string MasteryRootUrl = "/api/lol/static-data/{0}/v1.2/mastery";
+        private const string MasteriesUrl = "masteries";
+        private const string MasterbyIdUrl = "masteries/{0}";
```
Comment (BenFradet) [inline RiotSharp/StaticRiotApi.cs]: `MasteryByIdUrl`

File: RiotSharp/StaticRiotApi.cs [hunk 12]
```diff
@@ -24,46 +24,57 @@
...
-        private const string SummonerSpellsCacheKey = "spells";
-        private const string SummonerSpellCacheKey = "spell";
+        private const string SummonerSpellsUrl = "summoner-spells";
+        private const string SummonerSpellsCacheKey = "summoner-spells";
+        private const string SummonerSpellByIdUrl = "summoner-spells/{0}";
+        private const string SummonerSpellByIdCacheKey = "summoner-spell";
 
-        private const string VersionRootUrl = "/api/lol/static-data/{0}/v1.2/versions";
+        private const string VersionUrl = "versions";
         private const string VersionCacheKey = "versions";
```
Comment (BenFradet) [inline RiotSharp/StaticRiotApi.cs]: Should be `Versions` to be consistent

File: RiotSharp/StaticRiotApi.cs [hunk 13]
```diff
@@ -542,8 +544,7 @@ public List<MapStatic> GetMaps(Region region, Language language = Language.en_US
                 else
                 {
                     var json = requester.CreateGetRequest(
-                        string.Format(MasteryRootUrl, region.ToString()) + string.Format(IdUrl, masteryId),
-                        RootDomain,
+                        StaticDataRootUrl + string.Format(MasterbyIdUrl, masteryId), region,
```
Comment (BenFradet) [inline RiotSharp/StaticRiotApi.cs]: `MasteryByIdUrl`

File: RiotSharp/StaticRiotApi.cs [hunk 14]
```diff
@@ -562,19 +563,19 @@ public List<MapStatic> GetMaps(Region region, Language language = Language.en_US
...
             if (listWrapper != null && listWrapper.Language == language && listWrapper.MasteryData == masteryData)
             {
-                return listWrapper.MasteryListStatic.Masteries.ContainsKey(masteryId) ? listWrapper.MasteryListStatic.Masteries[masteryId] : null;
+                return listWrapper.MasteryListStatic.Masteries.ContainsKey(masteryId)
+                    ? listWrapper.MasteryListStatic.Masteries[masteryId] : null;
             }
             var json = await requester.CreateGetRequestAsync(
-                string.Format(MasteryRootUrl, region.ToString()) + string.Format(IdUrl, masteryId.ToString()),
-                RootDomain,
+                StaticDataRootUrl + string.Format(MasterbyIdUrl, masteryId), region,
```
Comment (BenFradet) [inline RiotSharp/StaticRiotApi.cs]: same

File: RiotSharpTest/App.config [hunk 15]
```diff
@@ -45,7 +45,7 @@
       <add key="TournamentPickType" value="TournamentDraft"/>
       <add key="TournamentMapType" value="SummonersRift"/>
       <!--CommonTest-->
-      <add key="ApiKey" value="YOUR_API_KEY"/>
+      <add key="ApiKey" value="486c4c93-7822-4184-a89c-e70773d4df55"/>
```
Comment (BenFradet) [inline RiotSharpTest/App.config]: !!!

```bash
git update-index --assume-unchanged RiotSharpTest/App.config
```

Comment (JanOuborny) [inline RiotSharpTest/App.config]: Removed and regenerated.

File: RiotSharp/StaticRiotApi.cs [hunk 16]
```diff
@@ -24,47 +24,58 @@
...
         private const string LanguagesCacheKey = "languages";
 
-        private const string MapRootUrl = "/api/lol/static-data/{0}/v1.2/map";
-        private const string MapCacheKey = "map";
+        private const string MapsUrl = "maps";
+        private const string MapsCacheKey = "maps";
 
-        private const string MasteryRootUrl = "/api/lol/static-data/{0}/v1.2/mastery";
+        private const string MasteriesUrl = "masteries";
+        private const string MasterByIdUrl = "masteries/{0}";
```
Comment (BenFradet) [inline RiotSharp/StaticRiotApi.cs]: `Master y ByIdUrl` :smile:

Comment (JanOuborny) [inline RiotSharp/StaticRiotApi.cs]: Am I not the master of ids?👑 🆔

**Comment by BenFradet** [comment]
@JanOuborny I think you'll need to rebase and we're good to go :+1:

**Pull request merged**